### DISCUSSION TITLE
React Query for flutter.

### DISCUSSION THREAD

**Discussion by 41y08h** [TanStack/query#3475] [Ideas] [answered]
Hey guys, I had started using react-query when my hate for redux was at its peak, and using it makes me comfortable so much so that **I want to port react-query to a dart package to be used in flutter**. I can't find anything like react-query or close to react-query for flutter so I want to build it myself, I'm familiar with react, react-query, and flutter/dart.

Therefore I want to know how react-query is implemented. Currently, I'm trying to understand this by looking into the codebase.
Any help would be appreciated.

Thanks :)

**Comment by TkDodo** [comment, marked as answer]
skip the `react` directory and have a look at the `core`. I can help if you have specific questions.

**Reply by 41y08h** [reply]
I've looked over things like query cache, query client, notifications, etc. I just want to know the general flow or in other words how they are connected together. Something like a query is created then it is pushed to the cache, a timer/background/worker process runs that refreshes the stale data, and then ... Thanks for giving your attention to this idea👍

**Reply by 41y08h** [reply]
Hey @TkDodo, I've got a little question. Is `QueryClient` class the bridge between core react-query and the react part of it?
In other words, Is `QueryClient` "the" class that the react part uses to interact with the core?

**Reply by TkDodo** [reply]
hmm, not exactly. If you take a look at how `useQuery` is implemented, it creates a `QueryObserver`. That is pretty much the class that lets you subscribe to a cache key and will inform you if something has changed.

The `QueryClient` is just a "vessel" that contains the `QueryCache` and the `MutationCache`, and provides functions to easily interact with it, like `queryClient.setQueryData` or `queryClient.fetchQuery`. It is also what you'll get with `useQueryClient` - it's just a wrapper around `useContext` from react.

So the structure is essentially:
- one QueryClient, that holds one QueryCache and one MutationCache
- multiple QueryObservers that are subscribed to a specific key of the cache
- when data is fetched, we put it in the cache, then notify all relevant observers about the change.
- when an observer mounts, it can also potentially trigger a fetch (e.g. if there is no data yet, the observer is not disabled etc ...)

**Reply by 41y08h** [reply]
Hey @TkDodo, let's suppose that react query was made to not have to provide a `QueryClient` and wrap the whole app in a `QueryClientProvider`, then would it have been possible to make react-query just only do the "work"? I just want to know what is "the main" reason we have to provide the `QueryClient` ourselves. From my understanding, its main reason is to allow the user to configure it.

Is `QueryClientProvider` anyway involved in notifying the components to rerender like the ContextAPI? From what I understand by looking at the code, a `QueryObserver` that is created in every `useQuery` hook instance is responsible for rerendering whenever the state changes.

**Reply by 41y08h** [reply]
I know you told me about this above, but I'm just confirming my views.

**Reply by TkDodo** [reply]
no, the Provider is react specific. You don't need it in other languages probably.

**Reply by 41y08h** [reply]
Here it is, I've developed a dart package, [FQuery](https://pub.dev/packages/fquery). It's still in beta version, a stable one will be released soon.

**Reply by JCKodel** [reply]
Also https://pub.dev/packages/query_stack (no need for Flutter hooks)
